src/c4/C4_MPI.cc

```diff
@@ -186,6 +186,36 @@ void wait_all(unsigned count, C4_Req *requests) {
   return;
 }
 
+//----------------------------------------------------------------------------//
+std::vector<int> wait_all_with_source(unsigned count, C4_Req *requests) {
+
+  // Nothing to do if count is zero.
+  if (count == 0)
+    return std::vector<int>();
+
+  // Return value -- rank IDs for all message sources.
+  std::vector<int> msg_sources(count);
+
+  std::vector<MPI_Request> array_of_requests(count);
+  // This is needed to obtain the source rank for each message:
+  std::vector<MPI_Status> array_of_statuses(count);
+  for (unsigned i = 0; i < count; ++i) {
```

Comment:
Unless there is something I'm missing-- no, I don't think I can. I need to loop over the requests in order, because I'm stashing the result for requests[i] in array_of_requests[i]. Is there a way to do this with range-based loops?